lib/elixir/lib/partition_supervisor.ex

```diff
@@ -302,15 +333,27 @@ defmodule PartitionSupervisor do
   ## Via callbacks
 
   @doc false
-  def whereis_name({name, key}) when is_atom(name) do
+  def whereis_name({name, key}) do
     partitions = partitions(name)
 
     partition =
       if is_integer(key), do: rem(abs(key), partitions), else: :erlang.phash2(key, partitions)
 
+    whereis_name(name, partition)
+  end
+
+  defp whereis_name(name, partition) when is_atom(name) do
     :ets.lookup_element(name, partition, 2)
   end
 
+  defp whereis_name(name, partition) when is_tuple(name) do
+    supervisor_pid = GenServer.whereis(name)
+
+    @registry
+    |> Registry.select([{{{:"$1", partition}, supervisor_pid, :_}, [], [:"$1"]}])
```

Comment:
Using a select seemed more straightforward than a lookup + find. The `supervisor_pid` gets us all of the partitions, and we need to extract that specific partition somehow.

Reply:
You should prefer lookups because lookups are guaranteed to be optimized. For example, this one is doing a full table traversal (I think). :) I will push a tiny change soon.